### ISSUE TITLE
Version 1.2.0 breaks support for Workflow/Pipeline jobs

### ISSUE THREAD

**Issue by omehegan** [jenkinsci/gitlab-plugin#271]
I should have tested this refactor with my usage of Pipeline, I could have caught this issue before we released. The problem is that we have used AbstractProject and AbstractBuild in many places. To be compatible with Pipeline as well as traditional freestyle jobs, we need to use Job<> instead of AbstractProject and Run<> instead of AbstractBuild. There is a more detailed guide to compatibility here:

https://github.com/jenkinsci/pipeline-plugin/blob/master/DEVGUIDE.md

I'll look at fixing this, but it's probably beyond my ability. We should also add a unit test that will verify compatibility going forward.

**Comment by omehegan** [comment]
See also the previous changes we made to support this:

https://github.com/jenkinsci/gitlab-plugin/pull/100
https://github.com/jenkinsci/gitlab-plugin/pull/206

**Comment by coder-hugo** [comment]
@omehegan  can you just try if this 1640a7e04a396a8d3e7a954c71e86dbc634bfef0 fixes the issue. I replaced AbstractProject and AbstractBuild with Job and Run wherever possible but there are places where it's not possible due to the inheritance hierarchy.

**Comment by omehegan** [comment]
I will take a look. Perhaps @jglick or someone can suggest an approach to take in the other places.

**Comment by omehegan** [comment]
@coder-hugo well, your commit will allow a build to trigger, so that's an improvement! Let me try out some other things and see if I find any issues. Thanks!

**Comment by omehegan** [comment]
Merge Request builds are broken for me in both Pipeline and Freestyle jobs, I think this is a separate bug.

**Comment by omehegan** [comment]
When I combine your https://github.com/jenkinsci/gitlab-plugin/commit/1640a7e04a396a8d3e7a954c71e86dbc634bfef0 patch with my changes for https://github.com/jenkinsci/gitlab-plugin/commit/f737542f8ae6b04c58d567ccc671eb32e391f565 I am at the point where a Pipeline job will be triggered when I create an MR. However, none of the appropriate job parameters, such as gitlabSourceBranch, are set, so I can't successfully check out code.

I think this is because setUpEnvironment, here: https://github.com/jenkinsci/gitlab-plugin/blob/feature/workflow-plugin-support/src/main/java/com/dabsquared/gitlabjenkins/listener/GitLabEnvironmentRunListener.java#L21 is using AbstractBuild. I am told that this simply won't work with Pipeline, so we either need to refactor this or have a special case to set the parameters when the job type is Pipeline.

**Comment by omehegan** [comment]
I have just posted https://groups.google.com/forum/#!topic/jenkinsci-dev/Z227uLfB81w to ask for help from the Jenkins community.

**Comment by jglick** [comment]
Use `EnvironmentContributor` rather than `RunListener.setUpEnvironment`.

**Comment by vekexasia** [comment]
I don't know if this is related but i'm getting this error when pushing to gitlab

```
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.kohsuke.stapler.Function$InstanceFunction.invoke(Function.java:320)
	at org.kohsuke.stapler.Function.bindAndInvoke(Function.java:163)
	at org.kohsuke.stapler.MetaClass$11.dispatch(MetaClass.java:378)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.MetaClass$11.dispatch(MetaClass.java:380)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:649)
	at org.kohsuke.stapler.Stapler.service(Stapler.java:238)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1669)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:135)
	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:126)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
	at hudson.security.UnwrapSecurityExceptionFilter.doFilter(UnwrapSecurityExceptionFilter.java:51)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:117)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.providers.anonymous.AnonymousProcessingFilter.doFilter(AnonymousProcessingFilter.java:125)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.rememberme.RememberMeProcessingFilter.doFilter(RememberMeProcessingFilter.java:135)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.AbstractProcessingFilter.doFilter(AbstractProcessingFilter.java:271)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.BasicHeaderProcessor.doFilter(BasicHeaderProcessor.java:93)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.context.HttpSessionContextIntegrationFilter.doFilter(HttpSessionContextIntegrationFilter.java:249)
	at hudson.security.HttpSessionContextIntegrationFilter2.doFilter(HttpSessionContextIntegrationFilter2.java:67)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:76)
	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:81)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:553)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at winstone.BoundedExecutorService$1.run(BoundedExecutorService.java:77)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
	at com.dabsquared.gitlabjenkins.webhook.ActionResolver.onPost(ActionResolver.java:93)
	at com.dabsquared.gitlabjenkins.webhook.ActionResolver.resolveAction(ActionResolver.java:53)
	at com.dabsquared.gitlabjenkins.webhook.ActionResolver.resolve(ActionResolver.java:47)
	at com.dabsquared.gitlabjenkins.webhook.GitLabWebHook.getDynamic(GitLabWebHook.java:44)
	... 65 more
```

**Comment by coder-hugo** [comment]
@jglick Before using the `RunListener` I used the `EnvironmentContributor` but IIRC this gets called to late and environment variables used in the GitSCM configuration aren't replaced by the corresponding values.

**Comment by omehegan** [comment]
@coder-hugo I have a feeling that we might need to use EnvironmentContributor AND RunListener, depending on which job is being used. I think I understand the issue you were having. In a Pipeline job, all of the SCM stuff happens after the job has started, as part of the Pipeline script, rather than being a part of the job configuration itself.

**Comment by omehegan** [comment]
@vekexasia this is not related, please file a separate issue.

**Comment by jglick** [comment]
> this gets called to late and environment variables used in the GitSCM configuration aren't replaced by the corresponding values

Sounds like a core bug if so. Do not recall the behavior offhand, would need to dig into sources. (Try adding a `Thread.dumpStack`.)

**Comment by coder-hugo** [comment]
I just tried it again using the `EnvironmentContributor` and now it's working. But we have updated also the minimum required versions of Jenkins and the git-plugin in the meantime. Maybe there was a bug in an early release of them.

**Comment by coder-hugo** [comment]
@omehegan I added this change to the branch feature/workflow-plugin-support. Can you try if your pipeline jobs are working with this.

**Comment by omehegan** [comment]
@coder-hugo unfortunately the Pipeline job still fails with:
`groovy.lang.MissingPropertyException: No such property: gitlabSourceBranch for class: WorkflowScript`

In other words the parameter is still not set. A freestyle job works OK. The log doesn't show any exceptions, the variables still just aren't getting set in this case.

**Comment by omehegan** [comment]
Oh, I see what the problem is. With this method, the variables are only accessible within the env[] map... or whatever it is. So I need to use env.gitlabSourceBranch.

**Comment by omehegan** [comment]
@coder-hugo OK, I think that fix is the best we can do. Per @jglick it is not currently possible to make all these values accessible as global Groovy variables, though he has a plan to support that in the future. Doesn't help us at the moment. With your change, freestyle jobs will be unaffected, but anyone with an existing Pipeline job is going to have to change their variable access to use env.

**Comment by coder-hugo** [comment]
@omehegan OK, til now I wasn't aware that the pipeline plugin handles build parameters and environment variables different as Jenkins itself makes no difference between them.
Maybe we can add a migration code that replaces all known gitlab-plugin variables within the groovy scripts by the corresponding environment variable during the upgrade of the plugin.

**Comment by omehegan** [comment]
@coder-hugo I talked to Jesse about this, but as I suspected there is no way we could perform this migration in users' Pipeline Groovy code. What we should do is release version 1.2.1 with your fix, and mark it as incompatible with prior versions using this approach: https://wiki.jenkins-ci.org/display/JENKINS/Marking+a+new+plugin+version+as+incompatible+with+older+versions  Then we can document what change people will need to make if they are using Pipeline. Given that Pipeline support is relatively new to the plugin, hopefully this will not impact too many people.

If this sounds OK, I propose we merge your change and release it today.
